### PR TITLE
expectedDamageFrames

### PR DESCRIPTION
**Pull request by DIProgan**
Attempt at boss health handling with expectedDamageFrames and weapon cooldownFrames (rate of fire)

### CODE REVIEW

File: enemies/bosses/main.json [hunk 1]
```diff
@@ -22,6 +22,7 @@
         ]
       },
       "hp": 18000,
+      "expectedDamageFrames": 300,
```
Comment (fdtaillefer) [inline enemies/bosses/main.json]: 18000 hp at 20 hp per hit, 4 shots per second, adds up to a 225-second fight, or 13500 frames. Assuming contact damage, you now logically need 2 etanks to beat Ceres Ridley.

File: enemies/bosses/main.json [hunk 2]
```diff
@@ -420,6 +428,7 @@
         ]
       },
       "hp": 13500,
+      "expectedDamageFrames": 1200,
```
Comment (fdtaillefer) [inline enemies/bosses/main.json]: Why would this fight take more damage if it lasts longer? You don't move and he can't hit you.

**Comment by fdtaillefer** [comment]
There's really very few bosses that you can reliably hit at the rate of fire of your weapon. The expected time between hits should also vary in some cases based on item loadout. Possibilities for mid-fight health farming aren't taken into account. Is this really the correct approach?

File: weapons/main.json [hunk 3]
```diff
@@ -181,11 +197,239 @@
...
+      "damage": 1200,
+      "cooldownFrames": 400,
+      "useRequires": ["Wave", "PowerBomb", "Charge"],
+      "shotRequires": [
+        {"ammo": {
+          "type": "PowerBomb",
+          "count": 1
+        }}
+      ],
+      "situational": false,
```
Comment (fdtaillefer) [inline weapons/main.json]: I think all SBAs should be considered situational.

File: region/tourian/main.json [hunk 4]
```diff
@@ -170,7 +170,11 @@
                       "excludedWeapons": [ "Super", "Missile" ]
                     }},
                     {"or": [
-                      {"metroidFrames": 300},
+                      {"enemyDamage": {
+                        "enemy": "Metroid",
+                        "type": "contact",
```
Comment (fdtaillefer) [inline region/tourian/main.json]: I wonder if that attack would be clearer with a more descriptive name, like drain16Frames or something?

**Comment by fdtaillefer** [comment]
For beam combo categories, charge is already handled.
For the others, I would lean towards `waveShot` / `iceshot` / `spazerShot` / `plasmaShot` categories. We can use those where appropriate in immunities and weaknesses, and for the weird cases like Evirs we can list them weapon by weapon (In the case of Evirs, I think that would mean 8 weapons to cover all the charged/uncharged ice combos that don't include plasma)

File: enemies/bosses/main.json (hunk 1 above)
Comment (DIProgan) [reply enemies/bosses/main.json]: I thought I changed it back to 0, doh

File: enemies/bosses/main.json (hunk 2 above)
Comment (DIProgan) [reply enemies/bosses/main.json]: My train of thought was something along the line of taking damage from thrown back supers that sometimes occur. Would you like to use 0?

Comment (fdtaillefer) [reply enemies/bosses/main.json]: I think the trajectory of those Supers may depend on your position, but yeah that's unclear.

File: weapons/main.json (hunk 3 above)
Comment (DIProgan) [reply weapons/main.json]: Basically dooming them to never get into logic considerations but maybe it's better that way for now.

File: region/tourian/main.json (hunk 4 above)
Comment (DIProgan) [reply region/tourian/main.json]: I thought of "drain" myself but felt that maybe it's better to keep a contact value for all enemies. I'm not exactly strongly opinionated on this.

**Comment by DIProgan** [comment]
> For beam combo categories, charge is already handled.
> For the others, I would lean towards `waveShot` / `iceshot` / `spazerShot` / `plasmaShot` categories. We can use those where appropriate in immunities and weaknesses, and for the weird cases like Evirs we can list them weapon by weapon (In the case of Evirs, I think that would mean 8 weapons to cover all the charged/uncharged ice combos that don't include plasma)

Weaknesses and strengths for combined beams seem to be nuts.
For example weapons that have 0.5x on Kihunter (red):
Ice
Spazer
Ice+Wave

Weapons that don't: 
Wave
Ice+Spazer
the rest

**Pull request merged**